repository-data/application/src/main/resources/hcm-config/configuration/update/Migration20180425UploadResourcesToS3.groovy

```diff
@@ -0,0 +1,158 @@
+
+        private DocumentHandle handle
+        private HashMap<Integer, S3ObjectMetadata> hashToS3Metadata
+
+        DocumentProcessor(handle) {
+            this.handle = handle
+            hashToS3Metadata = new HashMap<>()
+        }
+
+        boolean process() {
+            AtomicBoolean processed = new AtomicBoolean(false)
+
+            handle.getDocuments()
+                .values()
+                .each { variant ->
+
+                Node variantNode = variant.getNode(session)
+                new NodeIterable(variantNode.getNodes())
+                    .each { subNode -> convertNode(subNode, variantNode, processed) }
+            }
+
+            return processed.get()
+        }
+
+        private void convertNode(Node node, Node variantNode, AtomicBoolean processed) {
+            if (node.isNodeType(NODE_TYPE_OLD_ATTACHMENT)) {
+                Node resourceNode = node.getNode(NODE_NAME_RESOURCE)
+
+                if (resourceNode.hasProperty(PROPERTY_EXTERNAL_STORAGE_REFERENCE)) {
+                    log.info("node already migrated: " + resourceNode.getPath())
+                } else {
+                    log.info("attempting to migrate node: " + resourceNode.getPath())
+
+                    S3ObjectMetadata metadata = uploadToS3(resourceNode)
+                    setResourceProperties(resourceNode, metadata)
+
+                    // If we are on the published variant and it is live then publish the S3 resource
+                    if (variantNode.getProperty(HIPPOSTD_STATE).getString().equals(PUBLISHED) &&
+                        !variantNode.getProperty(HIPPOSTD_STATESUMMARY).getString().equals(STATE_NOT_PUBLISHED)) {
+                        log.info("publishing resource: " + metadata.reference)
+                        s3Connector.publishResource(metadata.reference)
+                    }
+
+                    processed.set(true)
+                }
+            }
+
+            new NodeIterable(node.getNodes())
+                .each { childNode -> convertNode(childNode, variantNode, processed) }
+        }
+
+        private S3ObjectMetadata uploadToS3(Node node) {
+            Binary binary = node.getProperty("jcr:data").getBinary()
+            InputStream stream = binary.getStream()
+
+            String fileName = node.getProperty(PROPERTY_FILE_NAME).getString()
+            String mimeType = node.getProperty(PROPERTY_MIME_TYPE).getString()
+
+            // We only want to upload a given file once, even if it is referenced by multiple variants
+            int hashCode = Objects.hash(streamComparer.getHashCode(stream), fileName, mimeType)
+            stream = binary.getStream()
+
+            S3ObjectMetadata metadata = hashToS3Metadata.get(hashCode)
+            if (metadata == null) {
+                log.info("About to upload to s3. file: $fileName mime: $mimeType hash: $hashCode")
+                metadata = s3Connector.scheduleUpload({ stream }, fileName, mimeType)
```

Comment:
I think we need to close the new instance of the stream after we're done reading it here; preferably from `finally` block.

I should've included the instruction in the description of the `scheduleUpload` method (I'll add it as part of my S3 logs story).

Reply:
Good spot, have updated